File: system-test/dotnet/Dsl.Core/Scenario/When/Steps/WhenPlaceOrder.cs
using Dsl.Core.Scenario.When.Steps.Base;
using Dsl.Port.When.Steps;
using Dsl.Core.Shared;
using Common;
using Driver.Adapter;
using Driver.Port.Shop.Dtos;
using Dsl.Core.Shop.UseCases;
using Optivem.Testing;
using static Dsl.Core.Gherkin.GherkinDefaults;

namespace Dsl.Core.Scenario.When.Steps;

public class PlaceOrder : BaseWhen<PlaceOrderResponse, PlaceOrderVerification>, IPlaceOrder
{
    private string? _orderNumber;
    private string? _sku;
    private string? _quantity;
    private string? _country;
    private string? _couponCode;

    public PlaceOrder(UseCaseDsl app, ScenarioDsl scenario, Func<Task> ensureGiven) : base(app, scenario, ensureGiven)
    {
        WithOrderNumber(DefaultOrderNumber);
        WithSku(DefaultSku);
        WithQuantity(DefaultQuantity);
        WithCountry(DefaultCountry);
        WithCouponCode(Empty);
    }

    public PlaceOrder WithOrderNumber(string? orderNumber)
    {
        _orderNumber = orderNumber;
        return this;
    }

    IPlaceOrder IPlaceOrder.WithOrderNumber(string? orderNumber) => WithOrderNumber(orderNumber);

    public PlaceOrder WithSku(string? sku)
    {
        _sku = sku;
        return this;
    }

    IPlaceOrder IPlaceOrder.WithSku(string? sku) => WithSku(sku);

    public PlaceOrder WithQuantity(string? quantity)
    {
        _quantity = quantity;
        return this;
    }

    IPlaceOrder IPlaceOrder.WithQuantity(string? quantity) => WithQuantity(quantity);

    public PlaceOrder WithQuantity(int quantity)
    {
        return WithQuantity(Converter.FromInteger(quantity));
    }

    IPlaceOrder IPlaceOrder.WithQuantity(int quantity) => WithQuantity(quantity);

    public PlaceOrder WithCountry(string? country)
    {
        _country = country;
        return this;
    }

    IPlaceOrder IPlaceOrder.WithCountry(string? country) => WithCountry(country);

    public PlaceOrder WithCouponCode(string? couponCode)
    {
        _couponCode = couponCode;
        return this;
    }

    IPlaceOrder IPlaceOrder.WithCouponCode(string? couponCode) => WithCouponCode(couponCode);

    public PlaceOrder WithCouponCode()
    {
        return WithCouponCode(DefaultCouponCode);
    }

    IPlaceOrder IPlaceOrder.WithCouponCode() => WithCouponCode();

    protected override async Task<ExecutionResult<PlaceOrderResponse, PlaceOrderVerification>> Execute(UseCaseDsl app)
    {
        var shop = await app.Shop(Channel);
        var result = await shop.PlaceOrder()
            .OrderNumber(_orderNumber)
            .Sku(_sku)
            .Quantity(_quantity)
            .Country(_country)
            .CouponCode(_couponCode)
            .Execute();

        return new ExecutionResultBuilder<PlaceOrderResponse, PlaceOrderVerification>(result)
            .OrderNumber(_orderNumber)
            .CouponCode(_couponCode)
            .Build();
    }
}

File: system-test/typescript/src/testkit/dsl/core/scenario/when/when-place-order.ts
import { DEFAULTS } from '../defaults.js';
import { UseCaseContext } from '../../shared/use-case-context.js';
import { AppContext } from '../app-context.js';
import { ScenarioContext } from '../scenario-context.js';
import { ThenResultStage } from '../then/then-place-order.js';

export class WhenPlaceOrder {
  private sku: string = DEFAULTS.SKU;
  private quantity: string | null = DEFAULTS.QUANTITY;
  private country: string = DEFAULTS.COUNTRY;
  private couponCode: string | null = null;

  constructor(
    private readonly app: AppContext,
    private readonly ctx: ScenarioContext,
    private readonly useCaseContext: UseCaseContext,
  ) {}

  withOrderNumber(_orderNumber: string): this {
    return this;
  }

  withSku(sku: string | null): this {
    this.sku = sku as string;
    return this;
  }

  withQuantity(quantity: string | number | null): this {
    this.quantity = quantity === null ? null : String(quantity);
    return this;
  }

  withCountry(country: string | null): this {
    this.country = country as string;
    return this;
  }

  withCouponCode(couponCode?: string | null): this {
    this.couponCode = couponCode === undefined ? DEFAULTS.COUPON_CODE : couponCode;
    return this;
  }

  then(): ThenResultStage {
    return new ThenResultStage(this.app, this.ctx, this.useCaseContext, this.sku, this.quantity, this.country, this.couponCode);
  }
}
